test/unit/xlog.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#define UNIT_TAP_COMPATIBLE 1
+#include "unit.h"
+#include "xlog.h"
+#include "xrow.h"
+#include "crc32.h"
+#include "random.h"
+#include "memory.h"
+#include "iproto_constants.h"
+
+/**
+ * Keep in sync with src/box/xlog.c!
+ */
+enum {
+	XLOG_READ_AHEAD_MIN = 128 * 1024,
+	XLOG_READ_AHEAD_MAX = 8 * 1024 * 1024,
+};
+
+/**
+ * Create a temporary directory, initialize it as xdir, and create a new xlog.
+ */
+static void
+create_xlog(struct xlog *xlog, char *dirname)
+{
+	if (mkdtemp(dirname) == NULL)
+		panic("mkdtemp() failed");
+
+	struct xdir xdir;
+	struct tt_uuid tt_uuid;
+	struct vclock vclock;
+	memset(&tt_uuid, 1, sizeof(tt_uuid));
+	memset(&vclock, 0, sizeof(vclock));
+
+	xdir_create(&xdir, dirname, XLOG, &tt_uuid, &xlog_opts_default);
+
+	if (xdir_create_xlog(&xdir, xlog, &vclock) < 0)
+		panic("xdir_create_xlog() failed");
+}
+
+/**
+ * Write a tuple to the xlog.
+ */
+static void
+write_tuple(struct xlog *xlog, const char *data, uint32_t size)
+{
+	static int64_t lsn;
+	struct request_replace_body body;
+	request_replace_body_create(&body, 0);
+
+	struct xrow_header row;
+	memset(&row, 0, sizeof(struct xrow_header));
+	row.lsn = ++lsn;
+	row.type = IPROTO_INSERT;
+	row.bodycnt = 2;
+	row.body[0].iov_base = &body;
+	row.body[0].iov_len = sizeof(body);
+	row.body[1].iov_base = (char *)data;
+	row.body[1].iov_len = size;
+
+	if (xlog_write_row(xlog, &row) < 0)
+		panic("xlog_write_row() failed");
+}
+
+/**
+ * Write 1 KB of random uncompressed data to the xlog. The compressed size
+ * is roughly the same due to the randomness of the data.
+ */
+static void
+write_1k(struct xlog *xlog)
+{
+	/* Each uint64_t is encoded as 9 bytes + the array header. */
+	const size_t num_elements = 114;
+	const size_t data_size = num_elements * 9 + 3;
+
+	char data[data_size];
+	char *data_end = mp_encode_array(data, num_elements);
+	for (size_t i = 0; i < num_elements; i++)
+		data_end = mp_encode_uint(data_end, xoshiro_random());
```

Comment:
Nit: You could encode `MP_BIN` to avoid all the computations above.

Reply:
Done